File: android/src/main/kotlin/com/yandex/appmetrica_plugin/Reporter.kt
/*
 * Version for Flutter
 * © 2022
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://yandex.com/legal/appmetrica_sdk_agreement/
 */

package com.yandex.appmetrica

import android.content.Context
import com.yandex.android.metrica.flutter.pigeon.Pigeon
import com.yandex.metrica.YandexMetrica

class Reporter(private val context: Context): Pigeon.ReporterPigeon {

    override fun sendEventsBuffer(apiKey: String) = YandexMetrica.getReporter(context, apiKey).sendEventsBuffer()

    override fun reportEvent(apiKey: String, eventName: String) =
        YandexMetrica.getReporter(context, apiKey).reportEvent(eventName)

    override fun reportEventWithJson(apiKey: String, eventName: String, attributesJson: Pigeon.StringPigeonWrapper) =
        YandexMetrica.getReporter(context, apiKey).reportEvent(eventName, attributesJson.stringPigeon)

    override fun reportError(apiKey: String, error: Pigeon.ErrorDetailsPigeon, message: Pigeon.StringPigeonWrapper) {
        YandexMetrica.getReporter(context, apiKey).pluginExtension.reportError(error.toNativeNotNull(), message.stringPigeon)
    }

    override fun reportErrorWithGroup(apiKey: String, groupId: String, error: Pigeon.ErrorDetailsPigeon, message: Pigeon.StringPigeonWrapper) {
        YandexMetrica.getReporter(context, apiKey).pluginExtension.reportError(groupId, message.stringPigeon, error.toNative())
    }

    override fun reportUnhandledException(apiKey: String, error: Pigeon.ErrorDetailsPigeon) {
        YandexMetrica.getReporter(context, apiKey).pluginExtension.reportUnhandledException(error.toNativeNotNull())
    }

    override fun resumeSession(apiKey: String) = YandexMetrica.getReporter(context, apiKey).resumeSession()

    override fun pauseSession(apiKey: String) = YandexMetrica.getReporter(context, apiKey).pauseSession()

    override fun setStatisticsSending(apiKey: String, enabled: Boolean) =
        YandexMetrica.getReporter(context, apiKey).setStatisticsSending(enabled)

    override fun setUserProfileID(apiKey: String, userProfileID: Pigeon.StringPigeonWrapper) =
        YandexMetrica.getReporter(context, apiKey).setUserProfileID(userProfileID.stringPigeon)

    override fun reportUserProfile(apiKey: String, userProfile: Pigeon.UserProfilePigeon) {
        YandexMetrica.getReporter(context, apiKey).reportUserProfile(userProfile.toNative())
    }

//    override fun putErrorEnvironmentValue(apiKey: String, key: String, value: Pigeon.StringPigeonWrapper) =
//        YandexMetrica.getReporter(context, apiKey).

    override fun reportRevenue(apiKey: String, revenue: Pigeon.RevenuePigeon) =
        YandexMetrica.getReporter(context, apiKey).reportRevenue(revenue.toNative())

    override fun reportECommerce(apiKey: String, event: Pigeon.ECommerceEventPigeon) {
        event.toNative()?.let(YandexMetrica.getReporter(context, apiKey)::reportECommerce)
    }
}

File: lib/src/reporter/reporter.dart
/*
 * Version for Flutter
 * © 2022
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://yandex.com/legal/appmetrica_sdk_agreement/
 */

import '../ecommerce_event.dart';
import '../error_description.dart';
import '../profile/attribute.dart';
import '../revenue.dart';

abstract class Reporter {

  /// Sends saved events from the buffer.
  Future<void> sendEventsBuffer();

  /// Sends a custom event message.
  Future<void> reportEvent(String eventName);

  /// Sends an event message in JSON format [attributesJson] 
  /// as a string and a short name or description of the event [eventName].
  Future<void> reportEventWithJson(String eventName, String? attributesJson);

  /// Sends an event as a set of attributes [attributes] Map
  /// and the short name or description of the event [eventName].
  Future<void> reportEventWithMap(String eventName, Map<String, Object>? attributes);

  /// Sends an error with the message [message] and the description [ErrorDescription]. 
  /// If there is no [errorDescription] description, the current stacktrace will be added automatically.
  Future<void> reportError({String? message, AppMetricaErrorDescription? errorDescription});

  /// Sends an error with its own identifier [groupId]. Errors in reports are grouped by it.
  Future<void> reportErrorWithGroup(String groupId, {AppMetricaErrorDescription? errorDescription, String? message});

  /// Sends an unhandled exception [error].
  Future<void> reportUnhandledException(AppMetricaErrorDescription error);

  /// Resumes the foreground session or creates a new one if the session timeout has expired.
  Future<void> resumeSession();

  /// Suspends the current foreground session.
  Future<void> pauseSession();

  /// Enables/disables sending statistics to the AppMetrica server.
  ///
  /// Disabling sending statistics for a reporter does not affect sending data from the main API key and other reporters. 
  /// But disabling sending data for the main API key stops sending statistics from all reporters.
  Future<void> setStatisticsSending(bool enabled);

  /// Sets the ID for the user profile.
  ///
  /// If Profile Id sending is not configured, predefined attributes are not displayed in the web interface.
  Future<void> setUserProfileID(String? userProfileID);

  /// Sends information about updating the user profile.
  Future<void> reportUserProfile(UserProfile userProfile);

  // Future<void> putErrorEnvironmentValue(String key, String? value);

  /// Sends the purchase information to the AppMetrica server.
  Future<void> reportRevenue(Revenue revenue);

  /// Sends ecommerce-event.
  Future<void> reportECommerce(ECommerceEvent event);
}
